### ISSUE TITLE
Gyro bias correction missing

### ISSUE THREAD

**Issue by mickeyxu** [CCNYRoboticsLab/imu_tools#11]
Hi, I've noticed one missing piece of code in your implementation compared to the Madgwick's sample code in his paper. 
Basically he also considered gyro drift (deg/s/s) in the parameter "zeta" to adjust the quaternion (when using mag readings). 

Is there a reason why this is not done here?

Thanks!

**Comment by idryanov** [comment]
I was working off the code he published online - do you have a link to the
code which uses zeta?

On Sat, Jun 8, 2013 at 2:04 AM, Mike notifications@github.com wrote:

> Hi, I've noticed one missing piece of code in your implementation compared
> to the Madgwick's sample code in his paper.
> Basically he also considered gyro drift (deg/s/s) in the parameter "zeta"
> to adjust the quaternion (when using mag readings).
> 
> Is there a reason why this is not done here?
> 
> Thanks!
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/ccny-ros-pkg/imu_tools/issues/11
> .

## 

Ivan Dryanovski
The Graduate Center, City University of New York (CUNY)
CCNY Robotics Lab
http://robotics.ccny.cuny.edu/People/Dryanovski

**Comment by mickeyxu** [comment]
Here's the report. http://www.x-io.co.uk/res/doc/madgwick_internal_report.pdf
At the end of it, the code with magnetic readings uses zeta for gyro correction.
Also, I don't quite get it why this correction is not done in his first example without mag values...

**Comment by mickeyxu** [comment]
Hi Ivan, Just wanted to check in if you have any ideas about this problem. 
Please keep me posted!
Thanks!

**Comment by idryanov** [comment]
Hi Mike,

Unfortunately I'm out of the lab for the next few months (doing an
internship at the moment) so I will not be able to add any features to the
package. If you want to implement this, I'd be happy to merge your pull
requests or give you commit access to the repo directly.

Ivan

On Mon, Jun 17, 2013 at 6:09 AM, Mike notifications@github.com wrote:

> Hi Ivan, Just wanted to check in if you have any ideas about this problem.
> Please keep me posted!
> Thanks!
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/ccny-ros-pkg/imu_tools/issues/11#issuecomment-19543472
> .

## 

Ivan Dryanovski
The Graduate Center, City University of New York (CUNY)
CCNY Robotics Lab
http://robotics.ccny.cuny.edu/People/Dryanovski

**Comment by mickeyxu** [comment]
Ivan,

I'd be happy to do a pull request. I'll need to learn how to do that, though...

**Comment by lorenzoriano** [comment]
Hi Mike,

have you already written the code to solve this issue? Is it in your GitHub fork? If so, it should be trivial to do a pull request. Please let me know what's your status and I'll try to help you.

**Comment by mickeyxu** [comment]
Lorenzo, Thanks for offering to help.
I think I figured out how to use the pull request:)

So, I've done the gyro drift correction part, and sent the pull request already.

Two notes about this part:
1. The value of zeta depends on the gyro characteristics. I tested it with the drift in yaw direction (z output), which is 0.003 (about 10 deg drift in 60 seconds), and it turned out to be okay. 
2. The gyro drift from MARG was almost negligible already. I personally think this gyro drift part should be implemented in the IMU filter that doesn't consider magnetometer. But Madgwick's paper says it is only applicable in MARG...
